### PR TITLE
Sparsity

### PR DESCRIPTION
**Pull request by fpicetti**
Added `optimization.sparsity` module with FISTA and SplitBregman algorithms.
Added Total Variation regularization example.

### CODE REVIEW

**Comment by mrava87** [comment]
Hi @fpicetti,
thanks a lot for your contribution, it looks great!

I am going to have a look in coming days and merge it to the master. 

From a quick skim through I see that the only missing part compared to the PyLops equivalent is the possibility to automatically infer `alpha` in the FISTA method. From a quick look it seems PyTorch is not there yet with power methods but it is worth keeping an eye on this Issue https://github.com/pytorch/pytorch/issues/8049.

Anything else that you know differs from the original PyLops implementations?

**Comment by fpicetti** [comment]
Hi @mrava87,

Yes, the FISTA method lacks of the `alpha` estimation. The example I added implements an Identity operator so it is straightforward to hard-code it.

Other differences: 

- my solvers have no `returninfo` option; I needed the objective function curves and I just returned all of them. As in this PR, FISTA returns a tuple of (`xinv`, `niter`, `cost`, `costdata`, `costreg`), while SB returns (`xinv`, `itn_out`, `cost`, `costdata`, `costregL1`, `costregL2`).
I can add the `returninfo` flag in order not to output the numpy arrays of the cost function terms.

- SplitBregman solver includes your `NormalEquationInversion` as linear solver based on CG, while in the original PyLops implementation it relies on the `RegularizedInversion` that is based on scipy's LSQR.

Thanks!

**Comment by mrava87** [comment]
Great :) this seems in line with what I could see when I skimmed through the code.

- Regarding the `returninfo`, ideally it would be good to be consistent with PyLops methods but we can always improve that later on. As long as we are on `0.x.x` version, breaking changes are allowed. Unfortunately we cannot change PyLops ones unless we bump to `2.x.x`, which is planned for some time in the future. In that case if there is anything that is useful to return in addition to what done now we could add it.

- For the `RegularizedInversion` vs `NormalEquationInversion`, I think they should behave about the same in this case. It should not be hard to include also `RegularizedInversion` in pylops-gpu, but again no need to do it in this PR. I can always be done later on.

- To the example: I agree, `alpha` can be found analytically for an `Identity` operator. As `alpha=1/max(eigs)` I would say it is `1` in this case. Can you explain your choice? Also, I have quickly tried to implement this example https://pylops.readthedocs.io/en/latest/gallery/plot_ista.html#sphx-glr-gallery-plot-ista-py in pylops-gpu (only using `FISTA`) and the results are in agreement with those from pylops. If you don't mind, I suggest we remove `FISTA` in the TV example, mostly because the signal that we want to denoise is clearly blocky (good for TV) but not sparse (bad for FISTA). It makes sense to me that FISTA doesn't do much there, while it works well in the https://pylops.readthedocs.io/en/latest/gallery/plot_ista.html#sphx-glr-gallery-plot-ista-py example where sparsity is a good assumption for both examples

I can commit some of these changes and then we can both have a final review before merge if its ok? :)



Any sorry for closing the PR, that was accidental!

**Comment by fpicetti** [comment]
Ok! I'll implement the `returninfo` option asap.
Totally agree with you about the CG/LSQR solvers for SB. Also, I will remove FISTA from the TV example: it was added just for the sake of completeness, as you said an identity regularizer in that case is not the best one.

**Comment by mrava87** [comment]
Great for adding ``returninfo`` :)

I made a small commit on top of the changes in the plot_tvreg.py: generating noise with numpy and ensuring the same seed gives the exact same noise; we can thus test solvers one to one. The L2 solver gives exactly the same solution, SplitBregman doesn't yet as we know due to the underlying solver used in the inner loop (I think...).

Do you want to add tests, or i can do it at some point during week-end.... after that I think this is good to be merged. Do you agree?

**Comment by fpicetti** [comment]
Yeah, thanks for the fixes :)
I can do some more tests on my own but it would be great to have your help on that. Week-end is ok! ;)

**Comment by mrava87** [comment]
Np :) I can write some unit tests, hopefully the same from pylops would work after some small adjustments. I will commit them for you to have a look once ready

**Comment by mrava87** [comment]
Hi,
thanks a lot for adding the power method.

I tried to use it while creating the tests and I noticed a strange behaviour that I cannot explain. It seems like that the convergence is unstable if I run it multiple times; this makes partially sense because the starting vector is randomly initiatized but for a large number of iterations we should expect convergence. I tried to look and found that wiki has a simply python code https://en.wikipedia.org/wiki/Power_iteration.

Here is my test:

```
import numpy as np
import torch
from pylops_gpu import MatrixMult

par = {'ny': 31, 'nx': 11, 'imag': 0, 'x0':False, 'dtype':'float64'} 

np.random.seed(0)
A = np.random.randn(par['ny'], par['nx']).astype(np.float32)
print(np.linalg.eig(A.T @ A)[0][0])

Aop = MatrixMult(torch.from_numpy(A).to(dev), device=dev)
print(_power_iteration(Aop.H @ Aop, 100)[0])
print(_power_iteration1(Aop.H @ Aop, 1000))
```

which gives:

```
65.76556
tensor(65.7656)
tensor(51.3805)
```

where `_power_iteration1` is your code that I simply wrapped in a function:

```
def _power_iteration1(Op, eigsiter=10, eigstol=0, dtype=torch.float32, device='cpu'):
    x = torch.rand(Op.shape[1]).type(dtype).to(device)
    x / x.norm()
    maxeig_old = torch.tensor(0.).type(dtype).to(device)
    
    for _ in range(eigsiter):
        y = Op * x
        maxeig = torch.abs(x.dot(y))
        x = y.clone()
        x /= x.norm()

        if maxeig - maxeig_old < eigstol * maxeig_old:
            break
        maxeig_old = maxeig.clone()
    return maxeig
```

and `_power_iteration` is the wiki one I converted to torch:

```
def _power_iteration(Op, niter=10, dtype=torch.float32, device='cpu'):
    # Choose a random vector decrease the chance that vector
    # is orthogonal to the eigenvector
    b_k = torch.rand(Op.shape[1]).type(dtype).to(device)

    for _ in range(niter):
        # Calculate the matrix-by-vector product Ab
        b_k1 = Op.matvec(b_k)

        # Calculate the norm
        b_k1_norm = torch.norm(b_k1)

        # Re-normalize the vector
        b_k = b_k1 / b_k1_norm
    # Compute the eigenvalue
    mu_k = torch.dot(b_k.T, Op.matvec(b_k)) / torch.norm(b_k)
    return mu_k, b_k
```

I noticed that you compute the eigenvalue at each iteration which allows having the stopping criterion but the algorithm is slightly different from the one on wiki, which reference did you follow?

**Comment by mrava87** [comment]
I think I found the problem.

This seems to make it work fine `if torch.abs(maxeig - maxeig_old) < eigstol * maxeig_old`. It was stopping as soon as the new eigenvalue estimate was smaller than the previous (difference is negative). I will do some more tests and if everything works fine will just move your power method into a private function and merge the PR. All the same tests from pylops for both FISTA and SplitBregman pass now :)

**Comment by mrava87** [comment]
I kept your method with the change I mentioned... they seem to both converge and have pretty much the same cost. I think its ready to merge. What do you think?

**Comment by fpicetti** [comment]
Hi,
Thanks for spotting out the problems!
For the power iteration, I don't have any reference. Indeed, my (buggy) implementation is very similar to the Wikipedia one but for the eigenvalue computation that enables the tolerance check (as an early stopping mechanism).
I tried your tests and also to me they work well. I think it's ready to merge, too!

**Comment by mrava87** [comment]
Very good :) thanks a lot for the contribution. 

I will add you to the list of contributors in the documentation and then merge :)
